src/components/heartrate/Ppg.cpp

```diff
@@ -2,98 +2,242 @@
-  if (dataIndex < 200)
-    return 0;
-
-  NRF_LOG_INFO("PREPROCESS, offset = %d", offset);
-  auto hr = ProcessHeartRate();
-  dataIndex = 0;
-  return hr;
+float Ppg::Mean(const float* signal, int start, int end) {
+  int n = 0;
+  float mean = 0.0f;
+  for (int i = start; i < end; i++) {
+    mean += signal[i];
+    n++;
+  }
+  if (n > 0) {
+    mean /= n;
+  }
+  return mean;
 }
-float Ppg::ProcessHeartRate() {
-  auto t0 = Trough(data.data(), dataIndex, 7, 48);
-  if (t0 < 0)
-    return 0;
 
-  float t1 = t0 * 2;
-  t1 = Trough(data.data(), dataIndex, t1 - 5, t1 + 5);
-  if (t1 < 0)
-    return 0;
-
-  float t2 = static_cast<int>(t1 * 3) / 2;
-  t2 = Trough(data.data(), dataIndex, t2 - 5, t2 + 5);
-  if (t2 < 0)
-    return 0;
+float Ppg::SignalToNoise(const float* signal, float max, int start, int end) {
+  float mean = Mean(signal, start, end);
+  return max / mean;
+}
+// Simple bandpass filter using exponential moving average
+void Ppg::Filter30to240(float* signal, int length) {
+  // From:
+  // https://www.norwegiancreations.com/2016/03/arduino-tutorial-simple-high-pass-band-pass-and-band-stop-filtering/
+  // 0.268 is ~0.5Hz and 0.816 is ~4Hz cutoff at 10Hz sampling
+  float expAlpha = 0.816f;
+  float expAvg;
+  int n, i;
+  for (n = 0; n < 4; n++) {
+    expAvg = signal[0];
+    for (i = 0; i < length; i++) {
+      expAvg = (expAlpha * signal[i]) + ((1 - expAlpha) * expAvg);
+      signal[i] = expAvg;
+    }
+  }
+  expAlpha = 0.268f;
+  for (n = 0; n < 4; n++) {
+    expAvg = signal[0];
+    for (i = 0; i < length; i++) {
+      expAvg = (expAlpha * signal[i]) + ((1 - expAlpha) * expAvg);
+      signal[i] -= expAvg;
+    }
+  }
+}
 
-  float t3 = static_cast<int>(t2 * 4) / 3;
-  t3 = Trough(data.data(), dataIndex, t3 - 4, t3 + 4);
-  if (t3 < 0)
-    return static_cast<int>(60 * 24 * 3) / static_cast<int>(t2);
+float Ppg::Max(const float* data, int start, int end) {
+  float max = 0.0f;
+  for (int i = start; i < end; i++) {
+    if (data[i] > max) {
+      max = data[i];
+    }
+  }
+  return max;
+}
 
-  return static_cast<int>(60 * 24 * 4) / static_cast<int>(t3);
+float Ppg::HeartRateAverage(float hr) {
+  avgIndex++;
+  avgIndex %= dataAverage.size();
+  dataAverage[avgIndex] = hr;
+  float avg = 0.0f;
+  float n = 0.0f, min = 300.0f, max = 0.0f;
+  for (const auto& value : dataAverage) {
+    if (value > 0.0f) {
+      avg += value;
+      if (value < min)
+        min = value;
+      if (value > max)
+        max = value;
+      n++;
+    }
+  }
+  if (n > 0) {
+    avg /= n;
+  } else {
+    avg = 0.0f;
+  }
+  return avg;
 }
 
-void Ppg::SetOffset(uint16_t offset) {
-  this->offset = offset;
-  dataIndex = 0;
+void Ppg::Detrend(float* signal, int size) {
+  float offset = signal[0];
+  float m = (signal[size - 1] - offset) / (size - 1);
+
+  for (int i = 0; i < size; i++) {
+    signal[i] -= (m * i + offset);
+  }
+  for (int i = 0; i < size - 1; i++) {
+    signal[i] = signal[i + 1] - signal[i];
+  }
 }
```

Comment:
Move functions that can be made static into an anonymous namespace. That way we know they don't cause side effects.

Reply:
Thanks for reviewing the code. The libs needed to be refactored to use float types rather than the original double types. This is done in order to take advantage of the FPU. Shall I move them into libs as is? Under a directory named signalprocessing or ?

Since we are working in frequency space we are primarily concerned with Nyquist and frequency resolution. The highest analyzed frequency is 4Hz (240bpm) so we need to acquire data at a minimum 8Hz. Frequency resolution is 1/TotalTime(seconds). I wanted to keep the total acquisition time to somewhere around 5 seconds for good UX. So I ended up at 10Hz with a 64 sample buffer resulting in a frequency resolution of 1/6.4s or 0.15625Hz, multiplied by 60 it is 9.375bpm. Yes that's poor resolution for pulse rate results! Thankfully the pulse rate frequency data has a nice Gaussian distribution and we can interpolate fairly accurate values. That coupled with pulse rate result averaging gives values that agree nicely with my off the shelf pulse oximeter. So we can certainly run at higher rates keeping in mind that we still need to acquire for at least 6.4s to keep the existing frequency resolution. It looks like it's taking ~20ms to run the existing algorithm, so the hard limit would be ~50Hz. Keep in mind that to keep the existing frequency resolution we would need to acquire for 6.4s resulting in a larger data buffer.

Reply:
Forgot to explain why it works. We can get better results using this method because the metrics for determining good versus bad pulse rate data are simpler. We can just look for a single peak above a defined threshold in a power spectrum that has low noise characteristics (simple max to average in this case) and assume that peak is the pulse. Then an output averaging ring buffer is used to store the good pulse rate values and also store failed pulse rates as zero. When the averaging buffer becomes all zeroes, reset the acquisition as if we had just pressed the "Start" button. So basically we just look at the values we think are good results. There are times one may experience incorrect readings but they should be short lived after physically remaining still and allowing good values to fill the buffer.

Reply:
Each lib should be in it's own directory. I suggest moving the libs to directories named Arduino-Interpolation and arduinoFFT according to the repo names where they're from.

Reply:
Libs moved.